File: lua/algorithms/bfs.lua
-- Breadth-First Search - Level-order graph traversal
--
-- Time Complexity: O(V + E)
-- Space Complexity: O(V)

local bfs = {}

function bfs.traverse(graph, start)
    local result = {}
    if not graph:hasVertex(start) then
        return result
    end

    local visited = {}
    local queue = { start }
    local front = 1
    visited[start] = true

    while front <= #queue do
        local vertex = queue[front]
        front = front + 1
        result[#result + 1] = vertex

        for _, neighbor in ipairs(graph:neighbors(vertex)) do
            if not visited[neighbor] then
                visited[neighbor] = true
                queue[#queue + 1] = neighbor
            end
        end
    end

    return result
end

function bfs.shortestPath(graph, start, target)
    if not graph:hasVertex(start) or not graph:hasVertex(target) then
        return {}
    end

    if start == target then
        return { start }
    end

    local visited = {}
    local parent = {}
    local queue = { start }
    local front = 1
    visited[start] = true

    while front <= #queue do
        local vertex = queue[front]
        front = front + 1

        if vertex == target then
            local path = {}
            local current = target
            while current ~= start do
                table.insert(path, 1, current)
                current = parent[current]
            end
            table.insert(path, 1, start)
            return path
        end

        for _, neighbor in ipairs(graph:neighbors(vertex)) do
            if not visited[neighbor] then
                visited[neighbor] = true
                parent[neighbor] = vertex
                queue[#queue + 1] = neighbor
            end
        end
    end

    return {}
end

function bfs.distances(graph, start)
    local result = {}
    if not graph:hasVertex(start) then
        return result
    end

    local queue = { start }
    local front = 1
    result[start] = 0

    while front <= #queue do
        local vertex = queue[front]
        front = front + 1
        local dist = result[vertex]

        for _, neighbor in ipairs(graph:neighbors(vertex)) do
            if result[neighbor] == nil then
                result[neighbor] = dist + 1
                queue[#queue + 1] = neighbor
            end
        end
    end

    return result
end

function bfs.pathExists(graph, start, target)
    if not graph:hasVertex(start) or not graph:hasVertex(target) then
        return false
    end

    if start == target then
        return true
    end

    local visited = {}
    local queue = { start }
    local front = 1
    visited[start] = true

    while front <= #queue do
        local vertex = queue[front]
        front = front + 1

        for _, neighbor in ipairs(graph:neighbors(vertex)) do
            if neighbor == target then
                return true
            end
            if not visited[neighbor] then
                visited[neighbor] = true
                queue[#queue + 1] = neighbor
            end
        end
    end

    return false
end

function bfs.connectedComponents(graph)
    local result = {}
    local visited = {}

    for _, vertex in ipairs(graph:vertices()) do
        if not visited[vertex] then
            local component = {}
            local queue = { vertex }
            local front = 1
            visited[vertex] = true

            while front <= #queue do
                local v = queue[front]
                front = front + 1
                component[#component + 1] = v

                for _, neighbor in ipairs(graph:neighbors(v)) do
                    if not visited[neighbor] then
                        visited[neighbor] = true
                        queue[#queue + 1] = neighbor
                    end
                end
            end

            result[#result + 1] = component
        end
    end

    return result
end

function bfs.isConnected(graph)
    if graph:isEmpty() then
        return true
    end

    local vertices = graph:vertices()
    if #vertices == 0 then
        return true
    end

    local traversed = bfs.traverse(graph, vertices[1])
    return #traversed == graph:vertexCount()
end

return bfs

File: julia/src/algorithms/bfs.jl
"""
Breadth-First Search - Level-order graph traversal

Time Complexity: O(V + E)
Space Complexity: O(V)
"""

"""
    bfs_traverse(g::Graph{T}, start::T) where T

Traverse graph starting from vertex using BFS.
Returns vertices in BFS order.
"""
function bfs_traverse(g::Graph{T}, start::T) where T
    result = T[]
    if !has_vertex(g, start)
        return result
    end

    visited = Set{T}()
    queue = T[start]
    push!(visited, start)

    while !isempty(queue)
        vertex = popfirst!(queue)
        push!(result, vertex)

        for neighbor in neighbors(g, vertex)
            if neighbor ∉ visited
                push!(visited, neighbor)
                push!(queue, neighbor)
            end
        end
    end

    result
end

"""
    shortest_path(g::Graph{T}, start::T, target::T) where T

Find shortest path between start and target.
Returns empty vector if no path exists.
"""
function shortest_path(g::Graph{T}, start::T, target::T) where T
    if !has_vertex(g, start) || !has_vertex(g, target)
        return T[]
    end

    if start == target
        return T[start]
    end

    visited = Set{T}()
    parent = Dict{T, T}()
    queue = T[start]
    push!(visited, start)

    while !isempty(queue)
        vertex = popfirst!(queue)

        if vertex == target
            # Reconstruct path
            path = T[]
            current = target
            while current != start
                pushfirst!(path, current)
                current = parent[current]
            end
            pushfirst!(path, start)
            return path
        end

        for neighbor in neighbors(g, vertex)
            if neighbor ∉ visited
                push!(visited, neighbor)
                parent[neighbor] = vertex
                push!(queue, neighbor)
            end
        end
    end

    T[] # No path found
end

"""
    distances(g::Graph{T}, start::T) where T

Calculate distances from start to all reachable vertices.
Returns Dict mapping vertex to distance.
"""
function distances(g::Graph{T}, start::T) where T
    result = Dict{T, Int}()
    if !has_vertex(g, start)
        return result
    end

    queue = T[start]
    result[start] = 0

    while !isempty(queue)
        vertex = popfirst!(queue)
        dist = result[vertex]

        for neighbor in neighbors(g, vertex)
            if !haskey(result, neighbor)
                result[neighbor] = dist + 1
                push!(queue, neighbor)
            end
        end
    end

    result
end

"""
    path_exists(g::Graph{T}, start::T, target::T) where T

Check if path exists between start and target.
"""
function path_exists(g::Graph{T}, start::T, target::T) where T
    if !has_vertex(g, start) || !has_vertex(g, target)
        return false
    end

    if start == target
        return true
    end

    visited = Set{T}()
    queue = T[start]
    push!(visited, start)

    while !isempty(queue)
        vertex = popfirst!(queue)

        for neighbor in neighbors(g, vertex)
            if neighbor == target
                return true
            end
            if neighbor ∉ visited
                push!(visited, neighbor)
                push!(queue, neighbor)
            end
        end
    end

    false
end

"""
    connected_components(g::Graph{T}) where T

Find all connected components in undirected graph.
"""
function connected_components(g::Graph{T}) where T
    result = Vector{Vector{T}}()
    visited = Set{T}()

    for vertex in vertices(g)
        if vertex ∉ visited
            component = T[]
            queue = T[vertex]
            push!(visited, vertex)

            while !isempty(queue)
                v = popfirst!(queue)
                push!(component, v)

                for neighbor in neighbors(g, v)
                    if neighbor ∉ visited
                        push!(visited, neighbor)
                        push!(queue, neighbor)
                    end
                end
            end

            push!(result, component)
        end
    end

    result
end

"""
    is_connected(g::Graph{T}) where T

Check if graph is connected (all vertices reachable from any vertex).
"""
function is_connected(g::Graph{T}) where T
    if isempty(g)
        return true
    end

    all_vertices = vertices(g)
    if isempty(all_vertices)
        return true
    end

    traversed = bfs_traverse(g, all_vertices[1])
    length(traversed) == vertex_count(g)
end
